Mega super hyper cleanup

pekkaklarck opened the pull request:
Cleanup to docs and code here and there. See commit messages of individual commits for details. Sorry for some of the commits grewing unfortunately large.

pekkaklarck committed:
Cleanup.

- Remove unnecessary coupling between Element and FormElement keywords.
- Enhance related error message.

pekkaklarck committed:
Quite big cleanup commit.

- Enhance documentation of ElementKeywords (#925).
- Add '.' at the end of error and log messages.
- Introduce `ContextsAware.find_elements` helper.
- Change default value of optional `message` from '' to None.
- Use `is_noney` instead of `is_falsy` when default value is None.
- General (and trivial) cleanup to code and tests.

aaltat commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -70,376 +69,380 @@ def current_frame_contains(self, text, loglevel='INFO'):
     @keyword
     def current_frame_should_not_contain(self, text, loglevel='INFO'):
-        """Verifies that current frame contains `text`.
+        """Verifies that current frame contains ``text``.
```
Should there be `not`  somewhere in the sentence?

pekkaklarck committed:
Enhance docs of different locator strategies.

Also added more XPath examples. This ought to cover enhancements
proposed by #940.

pekkaklarck committed:
Another big cleanup commit.

- Doc string to `find_element`. Mainly to give type hits to IDE but
  higher level APIs should get more docs in general.
- Introduced ElementNotFound error that `ElementFinder.find` uses.
- Move helpers to ContextAware and LibraryComponent to avoid
  unnecessary coupling between different library componets.
- Remove unnecessary helpers when `find_element(locator).method()`
  works as well.
- Rewriter internal logic with `Wait` keywords. This includes
  consistently handling non-existing elements.
- Make unregistering strategy that hasn't been registered an error.
- Consider `<input type="file">` text element consistently.
- General cleanup here and there.

pekkaklarck committed:
Small cleanup to logging

pekkaklarck committed:
Move frame related kws to own lib component

aaltat commented on src/SeleniumLibrary/keywords/frames.py:
```diff
@@ -0,0 +1,108 @@
+
+    @keyword
+    def current_frame_should_not_contain(self, text, loglevel='INFO'):
+        """Verifies that current frame contains ``text``.
```
I think doc string is missing text: not

aaltat commented on src/SeleniumLibrary/keywords/waiting.py:
```diff
@@ -249,30 +219,30 @@ def wait_until_element_does_not_contain(self, locator, text, timeout=None, error
+            try:
+                if condition():
+                    return
+            except ElementNotFound as err:
```
I was thinking that this code could resolve the StaleElementReferenceException if we would suppress also that exception. If it does, perhaps we should look us there other exceptions that could be also suppress.

aaltat commented:
Did only read to waiting object, but have few comments

pekkaklarck committed:
Consistent handling of enabled status.

"Element Should (Not) Be Enabled" and "Wait Until Element Is Enabled"
now all validate that the element is enabled and not readonly. Also
removed broken element type validation from the former keywords.

Fixes #958.

pekkaklarck committed:
Doc that elements considered enabled cannot be read-only

aaltat commented on src/SeleniumLibrary/errors.py:
```diff
@@ -0,0 +1,19 @@
+# limitations under the License.
+
+
+class ElementNotFound(Exception):
```
Having own exception is really good idea. 

I think in future it is possible that we have more exceptions and because it provides better information for the user. But for programmatic, I often have to need suppress all from specific class. Therefore it might be useful to have some base error class for SeleniumLibrary exceptions. Like: 

```
class BaseException(Exception):
    ROBOT_SUPPRESS_NAME = True


class ElementNotFound(BaseException):
    pass
```
Then in libraries which extend SL, I could do: 
```
try:
    self.sl.some_method(arg1, arg2...)
expect BaseException:
    do_something_else(ar3)
```

pekkaklarck replied:
Good idea. I'll add `SeleniumLibraryException` base class.

aaltat commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -18,428 +18,389 @@
+        using the log level specified with the optional ``loglevel``
+        argument. Valid log levels are ``DEBUG``, ``INFO`` (default),
+        ``WARN``, and ``NONE``. If the log level is ``NONE`` or below
+        the current active log level the source will not be logged.
```
It might be useful to mention that  text is verified from the DOM tree which sometimes may differ what is seen in the page. Example the case of the text might be different in the DOM, than what is seen in the page. Other valid example if the text is split to multiple elements, example there is link in the middle of the text, then keyword can not find the whole text. Instead the search must be divided between each element.

pekkaklarck replied:
Are you sure about the problem with finding text with elements in them? I just changed our test data to contain `This is the haystack and somewhere on this page is a <em>needle</em>.` and

    Page Should Contain    This is the haystack and somewhere on this page is a needle.

seems to work just fine.

aaltat replied:
I did have recollection that in some point it did not work, but I ques it does not hold water anymore.  But the first point is a valid one and sometimes causes questions in the user group or in the issue tracker.

aaltat commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -18,428 +18,389 @@
+                  % (count, locator))
 
     @keyword
     def page_should_not_contain(self, text, loglevel='INFO'):
```
The same comment that applies to `page_should_contain` method also applies to this method. Perhaps it applies even more, because it could leas to false positives more easily.

aaltat commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -18,428 +18,389 @@
+        self.find_element(locator).clear()
 
     @keyword
     def get_vertical_position(self, locator):
```
From code point of view, would it be goof idea to move this method after `get_horizontal_position` method?

pekkaklarck replied:
Yes, definitely. Should be done in a separate commit after this and other related PRs are merged. Moving stuff around has a high risk for conflicts.

aaltat replied:
Lets do the moving in another PR.

aaltat commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -474,68 +438,65 @@ def focus(self, locator):
-        Element can be moved on top of another element with `target`
-        argument.
+    def drag_and_drop(self, locator, target):
+        """Drags ``source`` element into ``target`` element.
```
change ``source`` to ``locator`` because the argument names have been changed.

pekkaklarck replied:
Good catch!

aaltat commented on src/SeleniumLibrary/keywords/element.py:
```diff
@@ -624,69 +586,68 @@ def press_key(self, locator, key):
-    def page_should_not_contain_link(self, locator, message='', loglevel='INFO'):
-        """Verifies image identified by `locator` is not found from current page.
+    def page_should_not_contain_link(self, locator, message=None, loglevel='INFO'):
+        """Verifies image identified by ``locator`` is not found from current page.
```
change "image" to "link"

aaltat commented on src/SeleniumLibrary/keywords/waiting.py:
```diff
@@ -174,42 +159,35 @@ def wait_until_element_is_not_visible(self, locator, timeout=None, error=None):
 
     @keyword
     def wait_until_element_is_enabled(self, locator, timeout=None, error=None):
         """Waits until element specified with `locator` is enabled.
```
\`locator ` to \``locator``

pekkaklarck replied:
I haven't gone through these docs yet.

aaltat replied:
No problem.

aaltat commented on src/SeleniumLibrary/keywords/waiting.py:
```diff
@@ -174,42 +159,35 @@ def wait_until_element_is_not_visible(self, locator, timeout=None, error=None):
 
-        Fails if `timeout` expires before the element is enabled. See
-        `introduction` for more information about `timeout` and its
+        Fails if `timeout` expires before the element is enabled. Element
```
\`timeout`  to  \``timeout``

aaltat commented on src/SeleniumLibrary/keywords/waiting.py:
```diff
@@ -174,42 +159,35 @@ def wait_until_element_is_not_visible(self, locator, timeout=None, error=None):
+        See `introduction` for more information about `timeout` and its
         default value.
 
         `error` can be used to override the default error message.
```
\`error` to \``error``

aaltat commented on src/SeleniumLibrary/locators/elementfinder.py:
```diff
@@ -119,11 +94,10 @@ def unregister(self, strategy_name):
-        else:
-            del self._strategies[strategy_name]
+        if strategy_name not in self._strategies:
+            raise RuntimeError("Cannot unregister the non-registered strategy '%s'."
```
Should there be issue about this change? It by a quick look feels backwards incompatible change.

pekkaklarck replied:
I guess so. Submitted #961.

aaltat reviewed (commented):
Did all the changes.

pekkaklarck committed:
Base class for SeLib exceptions

pekkaklarck committed:
Doc fixes based on review comments

pekkaklarck committed:
Test "Page Should Contain" with text spanning multiple elements.

pekkaklarck committed:
Little more cleanup.

- User `find_elements` instead of `find_element` when possible.
- Avoid `strategy=value` locator syntax in code.
- Enhance error messages when radio buttons not found.
- Cleanup `Choose File` test and test it with Firefox.
- TableElementFinder._locator to public class variable.
- Remove useless documentation.
